document-store/src/main/java/org/hypertrace/core/documentstore/mongo/query/parser/MongoLetClauseBuilder.java

```diff
@@ -0,0 +1,104 @@
+package org.hypertrace.core.documentstore.mongo.query.parser;
+
+import com.mongodb.BasicDBObject;
+import java.util.Map;
+import org.hypertrace.core.documentstore.expression.impl.AggregateExpression;
+import org.hypertrace.core.documentstore.expression.impl.AliasedIdentifierExpression;
+import org.hypertrace.core.documentstore.expression.impl.ArrayRelationalFilterExpression;
+import org.hypertrace.core.documentstore.expression.impl.ConstantExpression;
+import org.hypertrace.core.documentstore.expression.impl.DocumentArrayFilterExpression;
+import org.hypertrace.core.documentstore.expression.impl.FunctionExpression;
+import org.hypertrace.core.documentstore.expression.impl.IdentifierExpression;
+import org.hypertrace.core.documentstore.expression.impl.KeyExpression;
+import org.hypertrace.core.documentstore.expression.impl.LogicalExpression;
+import org.hypertrace.core.documentstore.expression.impl.RelationalExpression;
+import org.hypertrace.core.documentstore.expression.type.FilterTypeExpression;
+import org.hypertrace.core.documentstore.parser.FilterTypeExpressionVisitor;
+import org.hypertrace.core.documentstore.parser.SelectTypeExpressionVisitor;
+
+// Visitor for building the let clause in lookup stage from filter expressions
+class MongoLetClauseBuilder implements FilterTypeExpressionVisitor {
+
+  private final String subQueryAlias;
+
+  MongoLetClauseBuilder(String subQueryAlias) {
+    this.subQueryAlias = subQueryAlias;
+  }
+
+  @Override
+  public BasicDBObject visit(LogicalExpression expression) {
+    BasicDBObject letClause = new BasicDBObject();
+    for (FilterTypeExpression operand : expression.getOperands()) {
+      letClause.putAll((Map<String, Object>) operand.accept(this));
```

Comment:
We seem to be casting `BasicDBObject` into a `Map`. This could be confusing (probably compiler might be generating a warning).

Can we either convert everything to `BasicDBObject` or everything to `Map`?

Reply:
changed the return type to map